### ISSUE TITLE
Using callback for adjusting learning rate

### ISSUE THREAD

**Issue by bartlomiejcerek** [skorch-dev/skorch#505]
Hello! 

I have a question about using callbacks for adjusting the learning rate. 
Callbacks that I found in scorch documentation require an instance of an optimizer as input, which seemed problematic for me because I wasn't instantiating optimizer until the whole model initialization. 

After a few tries, I gave up and just created an instance of optimizer before the creation of the model and passed it as an argument but then I got an error that WarmRestartLR (the one I've been using) does not have set_param() function. 

So at this point, I am quite confused and I didn't see any example in the tutorial on how to use one of the ready callbacks from scratch.  If there are some examples could you please point me to it, or explain how to use those callbacks? 

Thank you in advance

**Comment by ottonemo** [comment]
From what I can gather you are trying to use a learning rate scheduler. The [Transfer Learning](https://github.com/skorch-dev/skorch/blob/master/notebooks/Transfer_Learning.ipynb) notebook shows you how to do that (albeit with a pre-defined policy but you can just define your own, or use `WarmRestartLR`). In essence, the following should work:

    net = NeuralNetClassifier(
        module,
        callbacks=[
            skorch.callbacks.LRScheduler(skorch.callbacks.WarmRestartLR),
        ]
    )

Note that you don't pass an *instance* of the policy but only the class. It will be instantiated later (and gets the optimizer it needs). You can pass additional arguments to the policy using named arguments passed to the `LRScheduler`, e.g. `LRScheduler(WarmRestartLR, base_period=20)`.

If you want to build a generic callback that changes the LR, please let me know.

**Comment by bartlomiejcerek** [comment]
Thank you very much, I missed that example! 

Yes, I would like to create a custom callback for changing LR rate, I guess I have to build it analogically to LRSheduler here? 
[https://github.com/skorch-dev/skorch/blob/master/skorch/callbacks/lr_scheduler.py](url)

I have one more question, I just tried lr scheduler as in the example:
`lrscheduler = LRScheduler(
    policy='StepLR', step_size=7, gamma=0.1)`

and I created other callback to monitor LR rate looking like that:
`class Monitor(Callback):
    def on_epoch_end(self, net, **kwargs):
        print(net.get_params()['lr'])`

and to my surprise value is not changing. Should I look for the updated value of LR somewhere else or my lr callback is not working?

**Comment by ottonemo** [comment]
> Yes, I would like to create a custom callback for changing LR rate, I guess I have to build it analogically to LRSheduler here?

The `LRScheduler` callback does not touch the optimizer, this is all done in the policies and I suggest that you implement your custom learning rate schedule as a policy if you can. If you have a very specific setup, you have to deal with the optimizer's param groups and all that. Minimal example to update the learning rate:

    for group in net.optimizer_.param_groups:
        group['lr'] = new_lr

> and to my surprise value is not changing. Should I look for the updated value of LR somewhere else or my lr callback is not working?

No, your callback is working fine. The problem with the `set_params`/`get_params` abstraction is that it is put on top of PyTorch and doesn't get feedback from PyTorch if an underlying value is changed. The scheduler policies do something like  `net.optimizer_.param_groups[param_group]['lr'] = 0.002` and since they come from PyTorch and have nothing to do with skorch they don't update the parameters that skorch holds which is why `net.get_params()['lr']` will not know about this change.

@thomasjpfan @BenjaminBossan do you think it would be a good idea to update this value internally? We have infrastructure in place to change the `lr` parameter during runtime but it could turn out to be difficult to track all the changes manually.

**Comment by BenjaminBossan** [comment]
Regarding changing `lr` during runtime, that would be a bad idea for several reasons:

* it's not straightforward
* there can be more than 1 learning rate
* if we do this, a call to `initialize()` would not actually reset the model since `lr` would have changed

As an alternative, what do you think about implementing an `lr_` attribute that, for convenience, collects all the learning rates from the param groups. Would that be possible and helpful?

**Comment by thomasjpfan** [comment]
Updating a parameter such as `lr`, would not be very scikit-learn like. 😅 

In the past, I have used a custom `LRRecorder` callback to place the learning rates of into history. (One would need to be careful when there is more than one learning rate)

**Comment by bartlomiejcerek** [comment]
Well for my simple usage obtaining lr by: 
`net.optimizer_.param_groups[param_group]['lr']`

Is good enough :) Thank you very much for your answers, it helped a lot! Everything is working now.
I find skorch super useful BTW !
